File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/WebServer.java
// Copyright (c) Microsoft Corporation
// All rights reserved. 
//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING 
// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 

package com.microsoft.frameworklauncher.webserver;

import com.microsoft.frameworklauncher.common.exceptions.AggregateException;
import com.microsoft.frameworklauncher.common.exceptions.NonTransientException;
import com.microsoft.frameworklauncher.common.exit.ExitStatusKey;
import com.microsoft.frameworklauncher.common.log.DefaultLogger;
import com.microsoft.frameworklauncher.common.model.FrameworkRequest;
import com.microsoft.frameworklauncher.common.model.FrameworkStatus;
import com.microsoft.frameworklauncher.common.model.LauncherConfiguration;
import com.microsoft.frameworklauncher.common.service.AbstractService;
import com.microsoft.frameworklauncher.common.service.StopStatus;
import com.microsoft.frameworklauncher.common.web.WebCommon;
import com.microsoft.frameworklauncher.zookeeperstore.ZookeeperStore;
import org.apache.hadoop.yarn.webapp.WebApps;

import java.util.List;
import java.util.Map;

// Forward Http Request to ZK Request and Return ZK Status.
public class WebServer extends AbstractService {
  private static final DefaultLogger LOGGER = new DefaultLogger(WebServer.class);

  private final LauncherConfiguration conf;

  /**
   * REGION SubServices
   */
  private final ZookeeperStore zkStore;
  private StatusManager statusManager;
  private RequestManager requestManager;

  /**
   * REGION AbstractService
   */
  public WebServer(LauncherConfiguration conf, ZookeeperStore zkStore) {
    super(WebServer.class.getName());
    this.conf = conf;
    this.zkStore = zkStore;
  }

  protected Boolean handleException(Exception e) {
    super.handleException(e);

    if (e instanceof NonTransientException) {
      LOGGER.logError(e,
          "NonTransientException occurred in %1$s. %1$s will be stopped.",
          serviceName);

      stop(new StopStatus(ExitStatusKey.LAUNCHER_INTERNAL_NON_TRANSIENT_ERROR.toInt(), true, null, e));
      return false;
    } else {
      LOGGER.logError(e,
          "Exception occurred in %1$s. It should be transient. Will restart %1$s inplace.",
          serviceName);

      // TODO: Only Restart WebServer instead of exit whole process and Restart by external system.
      stop(new StopStatus(ExitStatusKey.LAUNCHER_INTERNAL_UNKNOWN_ERROR.toInt(), false, null, e));
      return true;
    }
  }

  // No need to initialize for WebServer
  @Override
  protected void recover() throws Exception {
    super.recover();
    requestManager = new RequestManager(this, conf, zkStore);
    requestManager.start();

    // Run StatusManager depends on RequestManager
    statusManager = new StatusManager(this, conf, zkStore);
    statusManager.start();
  }

  @Override
  protected void run() throws Exception {
    super.run();

    // Here both RequestManager and StatusManager recover completed
    String bindAddress = WebCommon.getBindAddress(
        conf.getWebServerBindHost(), conf.getWebServerAddress());
    WebApps.$for("frameworklauncher", null, null, "ws")
        .at(bindAddress)
        .start(new LauncherWebApp(conf, statusManager, requestManager));

    LOGGER.logInfo("WebApp Started at %s", bindAddress);
  }

  // THREAD SAFE
  @Override
  public synchronized void stop(StopStatus stopStatus) {
    // Best Effort to stop Gracefully
    super.stop(stopStatus);

    AggregateException ae = new AggregateException();

    // Stop WebServer's SubServices
    try {
      if (zkStore != null) {
        zkStore.stop();
      }
    } catch (Exception e) {
      ae.addException(e);
    }

    if (ae.getExceptions().size() > 0) {
      LOGGER.logWarning(ae, "Failed to stop %s gracefully", serviceName);
    }

    LOGGER.logInfo("%s stopped", serviceName);
    System.exit(stopStatus.getCode());
  }


  /**
   * REGION Callbacks
   */
  // WebServer integrate and process all Callbacks from all its SubServices

  // Callbacks from SubServices
  public void onExceptionOccurred(Exception e) {
    LOGGER.logInfo(e, "onExceptionOccurred");

    // Handle SubService Exception ASAP
    handleException(e);
  }

  public void onCompletedFrameworkStatusesUpdated(
      Map<String, FrameworkStatus> completedFrameworkStatuses)
      throws Exception {
    requestManager.onCompletedFrameworkStatusesUpdated(completedFrameworkStatuses);
  }


  /**
   * REGION ReadInterface
   */
  public List<FrameworkRequest> getAllFrameworkRequests() throws Exception {
    return requestManager.getFrameworkRequests(null, null);
  }
}

File: subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/webserver/webserver.go
// MIT License
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE

package webserver

import (
	"context"
	"encoding/json"
	"fmt"
	si "github.com/microsoft/hivedscheduler/pkg/api"
	"github.com/microsoft/hivedscheduler/pkg/common"
	"github.com/microsoft/hivedscheduler/pkg/internal"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/klog"
	ei "k8s.io/kubernetes/pkg/scheduler/api"
	"net"
	"net/http"
	"strings"
	"time"
)

const (
	ComponentName = "webserver"
)

// The WebServer of HivedScheduler.
type WebServer struct {
	sConfig *si.Config

	// The backend http server
	server *http.Server

	// All paths to serve
	paths si.WebServerPaths

	// Scheduler Extender Callbacks from K8S Default Scheduler
	eHandlers internal.ExtenderHandlers

	// Scheduler Inspect Callbacks
	iHandlers internal.InspectHandlers
}

func NewWebServer(sConfig *si.Config,
	eHandlers internal.ExtenderHandlers,
	iHandlers internal.InspectHandlers) *WebServer {
	klog.Infof("Initializing " + ComponentName)

	ws := &WebServer{
		sConfig: sConfig,
		server: &http.Server{
			Addr: *sConfig.WebServerAddress,
		},
		paths:     si.WebServerPaths{Paths: []string{}},
		eHandlers: eHandlers,
		iHandlers: iHandlers,
	}

	ws.route(si.RootPath, ws.serve(ws.serveRootPath))
	ws.route(si.FilterPath, ws.serve(ws.serveFilterPath))
	ws.route(si.BindPath, ws.serve(ws.serveBindPath))
	ws.route(si.PreemptPath, ws.serve(ws.servePreemptPath))
	ws.route(si.AffinityGroupsPath, ws.serve(ws.serveAffinityGroups))
	return ws
}

func (ws *WebServer) route(path string, handler servePathHandler) {
	http.HandleFunc(path, handler)
	ws.paths.Paths = append(ws.paths.Paths, path)
}

func (ws *WebServer) AsyncRun(stopCh <-chan struct{}) <-chan struct{} {
	ln, err := net.Listen("tcp", *ws.sConfig.WebServerAddress)
	if err != nil {
		panic(fmt.Errorf(
			"Failed to listen on WebServerAddress: %v, %v",
			*ws.sConfig.WebServerAddress, err))
	}

	stoppedCh := make(chan struct{})
	go func() {
		defer close(stoppedCh)
		<-stopCh

		klog.Errorf("Stopping " + ComponentName)
		ctx, cancel := context.WithTimeout(context.Background(), 0)
		ws.server.Shutdown(ctx)
		cancel()
	}()

	go func() {
		// Blocking until error
		if err := ws.server.Serve(tcpKeepAliveListener{ln.(*net.TCPListener)}); err != nil {
			panic(fmt.Errorf("Error occurred while running WebServer: %v", err))
		}
	}()

	klog.Infof("Running " + ComponentName)

	return stoppedCh
}

// Copied from Golang net/http/server.go
type tcpKeepAliveListener struct {
	*net.TCPListener
}

func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
	tc, err := ln.AcceptTCP()
	if err != nil {
		return nil, err
	}
	tc.SetKeepAlive(true)
	tc.SetKeepAlivePeriod(3 * time.Minute)
	return tc, nil
}

// Error should be delivered by panic
type servePathHandler func(w http.ResponseWriter, r *http.Request)

func (ws *WebServer) serve(handler servePathHandler) servePathHandler {
	return func(w http.ResponseWriter, r *http.Request) {
		// The default error handling
		defer internal.HandleWebServerPanic(func(err *si.WebServerError) {
			// http.StatusOK is appended by default, so only need to explicitly specify
			// error statusCode.
			w.WriteHeader(err.Code)
			w.Write(common.ToJsonBytes(err.Message))
		})

		w.Header().Set("Content-Type", "application/json")
		handler(w, r)
	}
}

func (ws *WebServer) serveRootPath(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/" {
		panic(si.NewWebServerError(
			http.StatusNotFound,
			fmt.Sprintf("Path not found: %v", r.URL.Path)))
	}

	w.Write(common.ToJsonBytes(ws.paths))
}

func (ws *WebServer) serveFilterPath(w http.ResponseWriter, r *http.Request) {
	// recoverPanic to send error message to K8S
	defer internal.HandleWebServerPanic(func(err *si.WebServerError) {
		w.Write(common.ToJsonBytes(&ei.ExtenderFilterResult{Error: err.Error()}))
	})

	var args ei.ExtenderArgs
	if err := json.NewDecoder(r.Body).Decode(&args); err != nil {
		panic(internal.NewBadRequestError(fmt.Sprintf(
			"Failed to unmarshal web request body to ExtenderArgs: %v", err)))
	}

	// Args Defaulting
	if args.NodeNames == nil {
		args.NodeNames = &[]string{}
	}

	// Args Validation
	if args.Pod == nil {
		panic(internal.NewBadRequestError(fmt.Sprintf(
			"ExtenderArgs: Pod field should not be nil: %v",
			common.ToJson(args))))
	}

	w.Write(common.ToJsonBytes(ws.eHandlers.FilterHandler(args)))
}

func (ws *WebServer) serveBindPath(w http.ResponseWriter, r *http.Request) {
	// recoverPanic to send error message to K8S
	defer internal.HandleWebServerPanic(func(err *si.WebServerError) {
		w.Write(common.ToJsonBytes(&ei.ExtenderBindingResult{Error: err.Error()}))
	})

	var args ei.ExtenderBindingArgs
	if err := json.NewDecoder(r.Body).Decode(&args); err != nil {
		panic(internal.NewBadRequestError(fmt.Sprintf(
			"Failed to unmarshal web request body to ExtenderBindingArgs: %v", err)))
	}

	// Args Validation
	if args.PodNamespace == "" || args.PodName == "" ||
		args.PodUID == types.UID("") || args.Node == "" {
		panic(internal.NewBadRequestError(fmt.Sprintf(
			"ExtenderBindingArgs: All fields should not be empty: %v",
			common.ToJson(args))))
	}

	w.Write(common.ToJsonBytes(ws.eHandlers.BindHandler(args)))
}

func (ws *WebServer) servePreemptPath(w http.ResponseWriter, r *http.Request) {
	// Cannot send error message to K8S by ExtenderPreemptionResult, fallback to
	// the default error handling.

	var args ei.ExtenderPreemptionArgs
	if err := json.NewDecoder(r.Body).Decode(&args); err != nil {
		panic(internal.NewBadRequestError(fmt.Sprintf(
			"Failed to unmarshal web request body to ExtenderPreemptionArgs: %v", err)))
	}

	// Args Defaulting
	if args.NodeNameToMetaVictims == nil {
		args.NodeNameToMetaVictims = map[string]*ei.MetaVictims{}
	}

	// Args Validation
	if args.Pod == nil {
		panic(internal.NewBadRequestError(fmt.Sprintf(
			"ExtenderPreemptionArgs: Pod field should not be nil: %v",
			common.ToJson(args))))
	}

	w.Write(common.ToJsonBytes(ws.eHandlers.PreemptHandler(args)))
}

func (ws *WebServer) serveAffinityGroups(w http.ResponseWriter, r *http.Request) {
	name := strings.TrimPrefix(r.URL.Path, si.AffinityGroupsPath)
	if name == "" {
		if r.Method == http.MethodGet {
			w.Write(common.ToJsonBytes(ws.iHandlers.GetAffinityGroupsHandler()))
			return
		}
	} else {
		if r.Method == http.MethodGet {
			w.Write(common.ToJsonBytes(ws.iHandlers.GetAffinityGroupHandler(name)))
			return
		}
	}

	panic(internal.NewBadRequestError(fmt.Sprintf(
		"NotImplemented: %v: %v",
		r.Method, r.URL.Path)))
}
